### PR TITLE
[DPAS] Fix bugs in DPAS lowering.

### PR DESCRIPTION
**Pull request by chengjunlu**
This is second PR to support to lower the tt.dot to the DPAS operation.

A workable DPAS layout mapping with the threads_per_warp=16 for PVC. 
- Use the un-pack scalar type in converting layout from shared layout to DPAS layout.
- Pack the un-pack scalar type to i16/i32 of the operand A and B for the DPAS operation.
- Fix some bugs in the DPAS lowering.

There is one more PRs for:

- Add Intel optimization pass to transform the tt.dot with block layout to the DPAS layout.

### CODE REVIEW

**Comment by chengjunlu** [comment]
@etiotto @whitneywhtsang 
This PR is ready.
Please help to review the changes for lowering `tt.dot` to `DPAS`

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp [hunk 1]
```diff
@@ -248,13 +250,39 @@ struct ConvertLayoutOpConversion
...
+      //                      sub-group size 32
+      //               execution size = 16
+      // <------------------------------------------------------------->
+      // t0  t1  t2  t3  t4  t5  t6  t7  t8  t9  t10 t11 t12 t13 t14 t15       ^
+      // t16 t17 t18 t19 t20 t21 t22 t23 t24 t25 t26 t27 t28 t29 t30 t31       |
+      // .   .   .   .   .   .   .   .   .   .   .   .   .   .   .   .         | repeat count = 8
+      // .   .   .   .   .   .   .   .   .   .   .   .   .   .   .   .         |
+      // t0  t1  t2  t3  t4  t5  t6  t7  t8  t9  t10 t11 t12 t13 t14 t15       |
+      // t16 t17 t18 t19 t20 t21 t22 t23 t24 t25 t26 t27 t28 t29 t30 t31       v
+      // sizePerThreads = [2, 1]
```
Comment (chengjunlu) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:265]: Should be [4, 1]

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 2]
```diff
@@ -28,6 +43,133 @@ class DotOpDPASConversionHelper {
...
+    auto dTy = op.getD().getType().cast<RankedTensorType>();
+
+    // TODO: add more dpas supported data type.
+    if (dTy.getElementType().isIntOrIndex()) {
+      // Integer
+      if (dTy.getElementType().getIntOrFloatBitWidth() == 32) {
+        if (aTy.getElementType().getIntOrFloatBitWidth() == 8 &&
+            bTy.getElementType().getIntOrFloatBitWidth() == 8)
+          return dTy.getElementType().isSignedInteger()
+                     ? DPASEngineType::U32_U32_U8_U8
```
Comment (ienkovich) [inline lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp:61]: `S32_S32_S8_S8` here?

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 3]
```diff
@@ -28,6 +43,133 @@ class DotOpDPASConversionHelper {
...
+          return DPASEngineType::FP32_FP32_FP16_FP16;
+        if (aTy.getElementType().isBF16() && bTy.getElementType().isBF16())
+          return DPASEngineType::FP32_FP32_BF16_BF16;
+        if (aTy.getElementType().isF32() && bTy.getElementType().isF32() &&
+            op.getAllowTF32())
+          return DPASEngineType::FP32_FP32_TF32_TF32;
+      } else if (dTy.getElementType().isF16()) {
+        if (aTy.getElementType().isF16() && bTy.getElementType().isF16())
+          return DPASEngineType::FP16_FP16_FP16_FP16;
+      }
```
Comment (ienkovich) [inline lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp:77]: What about `BF16_BF16_BF16_BF16`?

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM.cpp [hunk 4]
```diff
@@ -77,11 +77,10 @@ struct DotOpConversion : public ConvertTritonGPUOpToLLVMPattern<triton::DotOp> {
           "Unsupported MMA kind found when converting DotOp to LLVM.");
     }
 
-    DpasEncodingAttr dpasLayout = D.getType()
-                                      .cast<RankedTensorType>()
-                                      .getEncoding()
-                                      .dyn_cast<DpasEncodingAttr>();
-    if (!isOuter && dpasLayout && supportDPAS(op)) {
```
[the remark this answers is not in the document]
Comment (chengjunlu) [reply lib/Conversion/TritonGPUToLLVM/DotOpToLLVM.cpp:84]: The new transform passes will check the validation to use the DPAS on 'tt.dot'.
If the DPAS layout exists, it means that the op supports the DPAS.

Remove `supportDPAS(op)` here to make it simple.

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp [hunk 5]
```diff
@@ -279,11 +280,10 @@ Value loadOperand(ConversionPatternRewriter &rewriter, Location loc,
 
   auto dpasLayout = encoding.getParent().cast<DpasEncodingAttr>();
   const SmallVector<unsigned> warpsPerCTA = dpasLayout.getWarpsPerCTA();
+  auto order = triton::gpu::getOrder(dpasLayout);
 
   auto tensorTy = tensor.getType().cast<RankedTensorType>();
   const ArrayRef<int64_t> tensorShape = tensorTy.getShape();
-  auto sharedLayout = tensorTy.getEncoding().cast<SharedEncodingAttr>();
```
[the remark this answers is not in the document]
Comment (chengjunlu) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp:285]: Done

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 6]
```diff
@@ -28,6 +43,133 @@ class DotOpDPASConversionHelper {
...
+      // floating.
+      if (dTy.getElementType().isF32()) {
+        if (aTy.getElementType().isF16() && bTy.getElementType().isF16())
+          return DPASEngineType::FP32_FP32_FP16_FP16;
+        if (aTy.getElementType().isBF16() && bTy.getElementType().isBF16())
+          return DPASEngineType::FP32_FP32_BF16_BF16;
+        if (aTy.getElementType().isF32() && bTy.getElementType().isF32() &&
+            op.getAllowTF32())
+          return DPASEngineType::FP32_FP32_TF32_TF32;
+      } else if (dTy.getElementType().isF16()) {
```
[the remark this answers is not in the document]
Comment (chengjunlu) [reply lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/DPAS.cpp:74]: Add BF16 base

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply lib/Conversion/TritonGPUToLLVM/DotOpToLLVM.cpp:84]: ok thanks for the explanation.

**Comment by chengjunlu** [comment]
> @chengjunlu a reminder to address code review comments in this PR so we can make progress and merge this important feature.

Yes.

The layout has not been settled yet. (The layout is different in software vectorization to pack the operand A to i16 or i32.)
I will update the PR along with the new layout changes.

**Review by chengjunlu — COMMENTED** [review]
looks good me. Thanks @whitneywhtsang for the help

**Comment by whitneywhtsang** [comment]
Depends on https://github.com/intel/llvm/pull/12821.

**Review by whitneywhtsang — APPROVED** [review]

**Review by etiotto — APPROVED** [review]

**Pull request merged**